src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -3534,14 +3537,81 @@ private static bool DirectoryInfoHasChildItems(DirectoryInfo directory)
                 }
                 else // Copy-Item local
                 {
+                    if (Context != null)
+                    {
+                        if (Context.ExecutionContext.SessionState.PSVariable.Get(SpecialVariables.ProgressPreferenceVarPath.ToString()).Value is ActionPreference progressPreference)
+                        {
+                            if (progressPreference == ActionPreference.Continue)
+                            {
+                                // we use a thread to gather the total files so that copying can start right away
+                                Thread t = new Thread(() =>
+                                {
+                                    GetTotalFiles(path, recurse);
+                                });
+                                t.Start();
+                                _copyStartTime = DateTime.Now;
+                            }
+                        }
+                    }
+
                     CopyItemLocalOrToSession(path, destinationPath, recurse, Force, null);
+                    if (_totalFiles > 0)
+                    {
+                        var progress = new ProgressRecord(COPY_FILE_ACTIVITY_ID, " ", " ");
+                        progress.RecordType = ProgressRecordType.Completed;
+                        WriteProgress(progress);
+                    }
                 }
             }
 
             _excludeMatcher.Clear();
             _excludeMatcher = null;
         }
 
+        private void GetTotalFiles(string path, bool recurse)
+        {
+            bool isContainer = IsItemContainer(path);
+
+            try
+            {
+                if (isContainer)
+                {
+                    var dir = new DirectoryInfo(path);
+                    foreach (var file in dir.EnumerateFiles())
+                    {
+                        Interlocked.Add(ref _totalFiles, 1);
+                        Interlocked.Add(ref _totalBytes, file.Length);
+                    }
+                }
+                else
+                {
+                    var file = new FileInfo(path);
+                    Interlocked.Add(ref _totalFiles, 1);
+                    Interlocked.Add(ref _totalBytes, file.Length);
+                }
+            }
+            catch
+            {
+                // ignore exception
+            }
+
+            if (recurse && isContainer)
+            {
+                try
+                {
+                    var dir = new DirectoryInfo(path);
+                    foreach (var subDir in dir.EnumerateDirectories())
+                    {
+                        GetTotalFiles(subDir.FullName, recurse);
```

Comment:
I wonder why we do recursion if there is `public static IEnumerable<string> EnumerateFiles(string path, string searchPattern, EnumerationOptions enumerationOptions)` and `EnumerationOptions` has `RecurseSubdirectories` and `IgnoreInaccessible`